File: bench/mini33_review.cpp
// Review fixes seen from the keyboard, on disposable disks only: the
// destination reread after a source-side failure, marks kept when nothing
// was tried, honest batch summaries, rename to the same name, a text file
// too large to edit whole, and DOS's $8D line ends in the editor.
//
// Checks do not stop the run, so one pass reports every finding that
// still reproduces; the exit code is the number of failed checks.
#include "Memory.h"
#include "M6502.h"
#include "DiskIICard.h"
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <string>
static std::string screen(Memory& m) {
    std::string s;
    for(int y=0;y<24;++y) { for(int x=0;x<40;++x) {
        unsigned char c=m.data()[0x400+(y&7)*128+(y>>3)*40+x]&127;
        if(c<32) c+=64; s+=c;
    } s+='\n'; } return s;
}
// the file rows of one panel: rows 2-20, 19 columns
static std::string panel(Memory& m,int side) {
    std::string s=screen(m),out;
    for(int y=2;y<21;++y) out+=s.substr(y*41+side*20,19)+"\n";
    return out;
}
static std::string mini_files(const char* suffix) {
    // the boot disk's file count, from the driver: the shipped disk may carry more than the four built files
    const char* n=getenv("MINI_FILES"); return std::string(n?n:"4")+suffix;
}
static void run(M6502& c,int cycles) { for(int n=0;n<cycles;) n+=c.run(1024); }
static int failures=0;
static void check(bool ok,const char* what,Memory& m) {
    if(ok) { printf("ok   %s\n",what); return; }
    ++failures; fprintf(stderr,"FAIL %s\n%s",what,screen(m).c_str());
}
int main(int argc,char** argv) {
    assert(argc==5);
    Memory m; m.setIIEMode(false);
    assert(m.loadAppleIIRom((std::string(argv[1])+"/roms/apple2p.rom").c_str()));
    auto card=std::make_unique<DiskIICard>(); auto *d=card.get();
    assert(d->loadBootRom(std::string(argv[1])+"/roms/disk2.rom"));
    assert(d->insertDisk(0,argv[2])); assert(d->insertDisk(1,argv[3]));
    d->setWriteBackEnabled(true); m.slotBus().plug(6,std::move(card));
    M6502 cpu(&m); m.setCpu(&cpu); cpu.setCpuMode(M6502::CpuMode::NMOS);
    m.clearRam(); m.resetSoftSwitches(); m.slotBus().reset(); cpu.hardReset();
    cpu.setProgramCounter(0xc600);
    run(cpu,180000000);
    if(screen(m).find(mini_files(" FILES").c_str())==std::string::npos) { fprintf(stderr,"no boot\n%s",screen(m).c_str()); return 99; }
    auto keys=[&](const char* s) { m.pasteRawKeys(s,strlen(s)); run(cpu,20000000); };
    auto wait=[&](const char* s) {
        for(int n=0;n<800 && screen(m).find(s)==std::string::npos;++n) run(cpu,1000000);
        return screen(m).find(s)!=std::string::npos;
    };
    auto settle=[&]() { run(cpu,40000000); };

    // 1. Tagged copy from drive 2: GOOD.A lands on drive 1, then BAD.B
    //    (a wrong sector count) fails while `drive` is the source.
    keys("\t/"); check(wait("BAD.B"),"drive 2 panel shows the copy fixture",m);
    keys("[ "); keys("K ");
    keys("C"); check(wait("COPY 2 MARKED"),"batch copy prompt",m);
    keys("Y"); check(wait("UNSUPPORTED"),"BAD.B refused",m); settle();
    check(panel(m,0).find("GOOD.A")!=std::string::npos,
          "F1: the destination panel is reread and shows GOOD.A",m);

    // 8. Both panels on drive 1: C tries nothing, so the mark stays.
    keys("\t="); keys(" ");
    auto flushes=d->getWriteFlushCount();
    keys("C"); check(wait("SELECT TWO DIFFERENT DRIVES"),"same-drive refusal",m); settle();
    check(screen(m).find("1 MARKED")!=std::string::npos,"F8: the mark survives a refused C",m);
    check(d->getWriteFlushCount()==flushes,"F8: nothing written",m);
    keys(" ");

    // The second fixture in drive 2, on the right panel.
    assert(d->insertDisk(1,argv[4]));
    keys("\t/"); check(wait("SAME.NAME"),"drive 2 panel shows the edit fixture",m);

    // 5. FULL.TXT: 32 sectors of text, the last byte not a NUL.
    keys("[K");
    keys("E");
    bool refused=wait("FILE TOO LARGE TO EDIT");
    check(refused,"F5: a text file that fills the area is not opened",m);
    if(!refused) { keys("\x1b"); keys("N"); }
    settle();

    // 7. Rename SAME.NAME to itself.
    keys("[KK");
    keys("R"); check(wait("RENAME: "),"rename prompt",m);
    keys("SAME.NAME"); keys("\r");
    check(wait("SAME NAME"),"F7: a same-name rename says so",m);
    check(screen(m).find("LOCKED - UNLOCK FIRST")==std::string::npos,
          "F7: not reported as a lock refusal",m);
    settle();

    // 4. TEXT.DOS is DOS text: ONE$8D TWO$8D, bit 7 set throughout.
    keys("[");
    keys("E"); check(wait("LEAVE"),"editor open",m);
    std::string s=screen(m);
    check(s.substr(0,3)=="ONE" && s.substr(41,3)=="TWO","F4: $8D ends a line on screen",m);
    keys("\x0a"); keys("X"); keys("\r"); keys("\x13");   // Ctrl-J: down a line; K is a letter now
    check(wait("NEW: "),"save asks for a name",m);
    keys("EDITED"); keys("\r");
    check(wait("CREATE TEXT FILE"),"create prompt",m);
    keys("Y"); check(wait("COPIED"),"edited text saved",m); settle();

    // 6. Lock batch: DEL.1 is unlocked, LOCKED.3 locked, so U-nlock both.
    keys("[KKK "); keys("KK ");
    keys("L"); check(wait("UNLOCK 2 MARKED"),"batch unlock prompt",m);
    keys("Y"); check(wait("1 UNLOCKED, 1 ALREADY SET"),"F6: lock batch summary",m); settle();

    // 6. Delete batch: DEL.1, DEL.2 and the still-locked KEEP.4.
    keys("[KKK "); keys("K "); keys("KK ");
    keys("D"); check(wait("DELETE 3 MARKED"),"batch delete prompt",m);
    keys("Y"); check(wait("2 DELETED, 1 LOCKED"),"F6: delete batch summary",m); settle();

    assert(d->flushPendingWrites());
    printf("%s: %d failed checks\n",failures?"FAIL":"PASS",failures);
    return failures;
}

File: bench/mini33_review.py
#!/usr/bin/env python3
"""Review fixes through the real UI, on temporary images; checks file bytes."""
import argparse
import os
import subprocess
import sys
import tempfile
from pathlib import Path
ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / 'tools'))
from mini33_fixture import make_disk, read_files, offset
from mkmini33 import MINI_VERSION
p = argparse.ArgumentParser(description=__doc__)
p.add_argument('--pom2-root', type=Path, required=True)
p.add_argument('--disk', type=Path, default=ROOT/f'dist/A2FILECMD-DOS3.3-{MINI_VERSION}.dsk')
a = p.parse_args()


def high(text):
    return bytes(c | 0x80 for c in text)


original = a.disk.read_bytes()
MINI_ENV = dict(os.environ, MINI_FILES=str(len(read_files(original))))   # the boot disk's file count, for the screen checks
copy_fixture = bytearray(make_disk([('GOOD.A', 0, b'GOOD DATA A'),
                                    ('BAD.B', 0, b'BAD DATA B')]))
copy_fixture[offset(17, 15) + 11 + 35 + 33] = 9     # a count the chain does not have
copy_fixture = bytes(copy_fixture)
edit_fixture = make_disk([('TEXT.DOS', 0, high(b'ONE\rTWO\r')),
                          ('FULL.TXT', 0, bytes([0xC1]) * 8192),
                          ('SAME.NAME', 0, b'SAME'),
                          ('DEL.1', 0, b'D1'), ('DEL.2', 0, b'D2'),
                          ('LOCKED.3', 0x80, b'L3'), ('KEEP.4', 0x80, b'K4')])
failures = []
with tempfile.TemporaryDirectory(prefix='mini33-review-') as tmp:
    d = Path(tmp)
    boot, first, second = d / 'boot.dsk', d / 'copy.dsk', d / 'edit.dsk'
    boot.write_bytes(original)
    first.write_bytes(copy_fixture)
    second.write_bytes(edit_fixture)
    subprocess.run([os.environ.get('CXX', 'c++'), '-std=c++17', '-O2',
                    *['-I' + str(a.pom2_root / s) for s in ('src', 'include', 'build/generated')],
                    str(ROOT / 'bench/mini33_review.cpp'),
                    str(a.pom2_root / 'build/libpom2_core_test.a'),
                    '-o', str(d / 'bench')], check=True)
    ui = subprocess.run([str(d / 'bench'), str(a.pom2_root), str(boot), str(first), str(second)],
                        env=MINI_ENV, timeout=600)
    if ui.returncode:
        failures.append(f'{ui.returncode} screen checks')

    def expect(ok, what):
        print(('ok   ' if ok else 'FAIL ') + what)
        if not ok:
            failures.append(what)

    expect(first.read_bytes() == copy_fixture, 'the copy source is untouched')
    before, after = read_files(original), read_files(boot.read_bytes())
    for name, f in before.items():
        expect(after.get(name, {}).get('data') == f['data'], f'boot {name} preserved')
    expect(after.get('GOOD.A', {}).get('data', b'').rstrip(b'\0') == b'GOOD DATA A',
           'GOOD.A copied byte for byte')
    expect('BAD.B' not in after, 'BAD.B not published')
    was, now = read_files(edit_fixture), read_files(second.read_bytes())
    for name in ('TEXT.DOS', 'FULL.TXT', 'SAME.NAME'):
        expect(now.get(name, {}).get('entry') == was[name]['entry'] and
               now[name]['data'] == was[name]['data'], f'{name} entry and bytes unchanged')
    expect(now.get('EDITED', {}).get('data', b'').rstrip(b'\0') == high(b'ONE\rX\rTWO\r'),
           'F4: typed text and RETURN follow the file ($D8, $8D)')
    expect('DEL.1' not in now and 'DEL.2' not in now, 'F6: DEL.1 and DEL.2 deleted')
    expect(now.get('LOCKED.3', {}).get('type') == 0 and now['LOCKED.3']['data'] == was['LOCKED.3']['data'],
           'F6: LOCKED.3 unlocked, bytes kept')
    expect(now.get('KEEP.4', {}).get('type') == 0x80 and now['KEEP.4']['data'] == was['KEEP.4']['data'],
           'F6: KEEP.4 still locked, not deleted')
    expect(a.disk.read_bytes() == original, 'the --disk image is unchanged')
if failures:
    print('FAIL:', '; '.join(failures))
    sys.exit(1)
print('PASS: dest reread, marks kept, batch summaries, same-name rename, full text refused, $8D editing')
